docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+On the other hand in the current system, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Very often, pods don’t use all the resources that they request, which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a system with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in another 4 pods of this type, but Kubernetes will reject new pods.
+
+In this design document, we propose **implementing two tiers of pods** - one in which we guarantee resources (top-tier pods) and the second wherein we provide resources if they’re available and at a slightly lower quality (best-effort pods). Adding a best-effort tier of pods will increase node’s utilization in a Kubernetes cluster (in Borg, supporting best-effort pods increased utilization by  about 20%). We also implement **a policy of automatically classifying pods that don’t provide limits as best-effort**. This provides much better resource guarantees to pods with limits specified.
+
+## QoS Functionality
+
+We will introduce two tiers of pods: top-tier pods and best-effort pods. We will not support multiple levels or priorities.
+
+Best-effort pods will be invisible to top-tier pods:
+- **Existence of best-effort pods should not prevent top tier pods from launching**. When deciding whether to launch a top-tier pod, we only look at resource limits of other top-tier pod. Suppose we have a machine with 4GB of memory. Top-tier pod P1 has a 2GB limit, and 4 best-effort pods have a 500MB limit each. Suppose the user launches a new top-tier pod P2 with a 1GB limit. The Kubelet will evict best-effort pods, in unspecified order, to make room for P2. P2 will then be admitted and allocated the requested resources.
+- **Allocation of excess compressible resources (e.g. CPU) to top-tier pods will be the same as if best-effort pods did not exist**. For example if top-tier pod P1 asks for minimum 20% CPU, and we also have best-effort pods B1 and B2 running. Suppose P1 needs more CPU (because of a huge number of requests coming in), then P1 will starve B1, B2 of CPU and use the whole 100% of the CPU (assuming there are no other top tier pods running).
+- **Allocation of non-compressible resources (e.g. memory) to top-tier pods will be the same as if best-effort pods did not exist**. For example consider a system with 4GB of memory, where top-tier pod P1 is using 500MB of memory (with a 1GB limit) and best-effort pods are using the other 3.5GB of memory. The best-effort pods will not prevent P1 from accessing more memory (assuming that P1 remains within its memory limits). So if P1 asks for 100MB more memory, P1 will get the memory it needs, and one of the other pods will be OOM killed. 
+
+Resource guarantees for best-effort pods:
+- Compressible resources
+  - For now, we are only implementing CPU limits.

```

Comment:
@AnanyaKumar - Imagine a situation where an operator is managing a cluster and applying ```ResourceQuota``` to each ```Namespace``` on the system.  Each ```Namespace``` is used by different user communities to run their app.  By the luck of the draw, a developer gets their 200m cpu container scheduled on a c4.8xlarge node with no other containers on it.  The operator was willing to do this because they wanted to accept the pod, and that was the only node available to meet the job.  The user's pod is able to do a ton of compute, and gets crazy good performance until other containers get scheduled to the node and things slow down over time.  The user calls the cluster operator to complain to either increase their quota, or say, 'what gives?  it used to be so much faster?'.  

To avoid calls like this in the future, the operator wants to impose a ceiling on the amount of cpu even if there is additional cpu available on the machine to just provide a more consistent measure of performance.  I recognize that right now we are not normalizing compute across different node profiles at this time, but in the case of a homogenous cluster of nodes, ```cpu-quota``` is of use.

In the past with OpenShift, we have used a combination of ```--cpu-shares``` and ```-cpu-quota``` to provide a consistent amount of performance independent of the amount of containers on a node, and a means to reconcile ```-cpu-quota``` across various types of node to attempt to provide some amount of consistent performance across what we called gear profiles (which align with various types of compute instances).

Generally, I think users of RHEL with cgroups would also want to be able to take advantage of the same underlying techniques they may have used when moving workloads to Kubernetes (see https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Resource_Management_Guide/sec-cpu.html)